File: src/lib/server/modules/juinjutsu.ts
import type { SearchResult, Manga, Chapter } from '../classes/interfaces';
import Module from '../classes/Module';
import axios from 'axios';
import { load as cheerioload } from 'cheerio';

class Juinjutsu extends Module {
	id = 'jj';
	name = 'JuinJutsu';

	search(query: string) {
		return new Promise(async (resolve: (value: SearchResult[]) => void) => {
			if (query.length < 2) {
				resolve([]);
				return;
			}

			const page1 = await axios.get('https://juinjutsureader.ovh/directory/1/');
			const parsed = cheerioload(page1.data);

			const pages_count = parseInt(
				parsed('.gbuttonright.btn.btn-primary').attr('href')?.split('directory/')[1].split('/')[0] +
					''
			);

			const pages = [];
			pages.push(page1.data);
			for (let i = 2; i <= pages_count; i++) {
				const p = await axios.get('https://juinjutsureader.ovh/directory/' + i + '/');
				pages.push(p.data);
			}

			const chs: { title: string; id: string }[] = [];
			pages.forEach((el) => {
				const $ = cheerioload(el);
				const caps = $('.series_element');
				caps.each((i, capEl) => {
					const cap = { title: '', id: '' };
					cap.title = $('.title', capEl).find('a').text();
					cap.id = $('.title', capEl).find('a').attr('href')?.split('series/')[1] + '';
					chs.push(cap);
				});
			});

			const results: { title: string; id: string }[] = [];
			chs.forEach((ele) => {
				if (ele.title.toLowerCase().includes(query.toLowerCase())) {
					results.push(ele);
				}
			});

			resolve(results);
		});
	}

	manga(mangaid: string) {
		return new Promise(async (resolve: (value: Manga) => void) => {
			const page = await axios.get('https://juinjutsureader.ovh/series/' + mangaid);
			const $ = cheerioload(page.data);

			const chapters: Chapter[] = [];
			const data = {
				img: '',
				synopsis: '',
				author: '',
				artist: '',
				sourceurl: 'https://juinjutsureader.ovh/series/' + mangaid,
				chapters: chapters
			};
			data.img = $('img.thumb').attr('src') + '';
			data.synopsis = $('.trama') ? $('.trama').text().substring(7) : '';
			data.author = $('.autore') ? $('.autore').text().substring(8) : '';
			data.artist = $('.artista') ? $('.artista').text().substring(9) : '';

			$('.element').each((i, el) => {
				const a = $(el).find('a');
				const chtitle = a.text();
				const chid = a.attr('href')?.split(mangaid).pop() + '';
				const chdate = psDate($(el).find('.meta_r').text());

				data.chapters.push({ title: chtitle, id: chid, date: chdate });
			});

			resolve(data);
		});
	}

	chapter(manga: string, id: string) {
		return new Promise(async (resolve: (value: string[]) => void) => {
			const page = await axios('https://juinjutsureader.ovh/read/' + manga + id);
			const json = JSON.parse(page.data.split('var pages = ')[1].split(';')[0]);

			const results: string[] = [];
			json.forEach((el: { url: string }) => {
				results.push(el.url);
			});
			resolve(results);
		});
	}
}

export default new Juinjutsu();

function psDate(date: string): Date {
	switch (date) {
		case 'Oggi':
			return new Date();
		case 'Ieri': {
			const now = new Date();
			now.setDate(now.getDate() - 1);
			return now;
		}
		default: {
			const dt = new Date();
			const p = date.split('.');
			dt.setFullYear(parseInt(p[0]));
			dt.setMonth(parseInt(p[1]) - 1);
			dt.setDate(parseInt(p[2]));
			return dt;
		}
	}
}

File: pkg/modules/juinjutsu/juinjutsu.go
package juinjutsu

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/Minettyx/FoolslideProxy/pkg/types"

	"github.com/PuerkitoBio/goquery"
)

type juinJutsu struct {
	baseUrl string
}

var Module = juinJutsu{
	baseUrl: "https://juinjutsureader.ovh/",
}

var _ types.Module = juinJutsu{}

func (c juinJutsu) Id() string {
	return "jj"
}
func (c juinJutsu) Name() string {
	return "JuinJutsu"
}
func (c juinJutsu) Flags() types.ModuleFlags {
	return types.ModuleFlags{}
}

func (c juinJutsu) Search(query string) ([]types.SearchResult, error) {
	res, err := http.PostForm(c.baseUrl+"search/", url.Values{
		"search": {query},
	})
	if err != nil {
		return nil, err
	}

	defer res.Body.Close()

	doc, err := goquery.NewDocumentFromReader(res.Body)
	if err != nil {
		return nil, err
	}

	results := []types.SearchResult{}

	doc.Find("div .series_element").Each(func(i int, s *goquery.Selection) {

		id, _ := s.Find("a").First().Attr("href")
		par := strings.Split(id, "series/")
		if len(par) < 2 {
			return
		}

		id = par[1]

		title := s.Find(".title > a").Text()

		image, _ := s.Find("img").First().Attr("src")

		results = append(results, types.SearchResult{
			Id:    id,
			Title: title,
			Image: image,
		})
	})

	return results, nil
}

func (c juinJutsu) Manga(id string) (*types.Manga, error) {
	res, err := http.Get(c.baseUrl + "series/" + id)
	if err != nil {
		return nil, err
	}

	defer res.Body.Close()

	doc, err := goquery.NewDocumentFromReader(res.Body)
	if err != nil {
		return nil, err
	}

	synopsis := doc.Find(".trama").Text()
	if len(synopsis) > 7 {
		synopsis = synopsis[7:]
	}

	author := doc.Find(".autore").Text()
	if len(author) > 8 {
		author = author[8:]
	}

	artist := doc.Find(".artista").Text()
	if len(artist) > 9 {
		artist = artist[9:]
	}

	img, _ := doc.Find(".thumb").Attr("src")

	chapters := []types.Chapter{}

	doc.Find(".element").Each(func(i int, s *goquery.Selection) {

		chtitle := s.Find("a").Text()
		chid, _ := s.Find("a").Attr("href")
		pts := strings.Split(chid, id)
		if len(pts) < 2 {
			return
		}
		chid = pts[1]

		datestr := s.Find(".meta_r").Text()
		date, err := parseDate(datestr)
		if err != nil {
			return
		}

		chapters = append(chapters, types.Chapter{
			Title: chtitle,
			Id:    chid,
			Date:  date,
		})
	})

	result := types.Manga{
		Synopsis:  synopsis,
		Author:    author,
		Artist:    artist,
		Img:       img,
		Sourceurl: "https://juinjutsureader.ovh/series/" + id,
		Chapters:  chapters,
	}

	return &result, nil
}

func (c juinJutsu) Chapter(manga, id string) ([]string, error) {
	res, err := http.Get(c.baseUrl + "read/" + manga + id)
	if err != nil {
		return nil, err
	}

	defer res.Body.Close()

	body, err := io.ReadAll(res.Body)
	if err != nil {
		return nil, err
	}

	pts := strings.Split(string(body), "var pages = ")
	if len(pts) < 2 {
		return nil, fmt.Errorf("Error getting pages for juinjutsu chapter")
	}

	jsonstr := strings.Split(pts[1], ";")[0]

	type pagesJson struct {
		Url string `json:"url"`
	}

	var jsonv []pagesJson

	err = json.Unmarshal([]byte(jsonstr), &jsonv)
	if err != nil {
		return nil, err
	}

	result := []string{}

	for _, v := range jsonv {
		result = append(result, v.Url)
	}

	return result, nil
}

func (c juinJutsu) Latest() ([]types.LatestResult, error) {
	return []types.LatestResult{}, nil
}

func (c juinJutsu) Popular() ([]types.PopularResult, error) {
	return []types.PopularResult{}, nil
}

func parseDate(date string) (time.Time, error) {
	switch date {
	case "Oggi":
		return time.Now(), nil
	case "Ieri":
		return time.Now().Add(time.Hour * -24), nil
	default:
		return time.Parse("2006.1.02", date)
	}
}
